bn-apps/cordapp-updates-distribution/corda-updates-app/src/main/kotlin/com/r3/businessnetworks/cordaupdates/app/bno/DatabaseService.kt

```diff
@@ -1,130 +1,136 @@
+            return dbObject
+        }
     }
 
-    init {
-        // Create table if it doesn't already exist.
-        val nativeQuery = """
-                create table if not exists $CORDAPP_VERSION_INFO_TABLE (
-                    $PARTY varchar(255) not null,
-                    $CORDAPP_GROUP varchar(255) not null,
-                    $CORDAPP_NAME varchar(255) not null,
-                    $CORDAPP_VERSION varchar(20) not null,
-                    $LAST_UPDATED bigint not null
-                );
-                create unique index if not exists party_group_name_unique_index on $CORDAPP_VERSION_INFO_TABLE($PARTY, $CORDAPP_GROUP, $CORDAPP_NAME);
-            """
-        val session : Connection = serviceHub.jdbcSession()
-        session.prepareStatement(nativeQuery).execute()
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "id", nullable = false)
+    var id : Long = 0
+
+    @Column(name = "party", nullable = false)
+    var party : String? = null
+
+    @Column(name = "cordapp_group", nullable = false)
+    var cordappGroup : String? = null
+
+    @Column(name = "cordapp_name", nullable = false)
+    var cordappName : String? = null
+
+    @Column(name = "cordapp_version", nullable = false)
+    var cordappVersion : String? = null
+
+    @Column(name = "last_updated", nullable = false)
+    var lastUpdated : Long? = null
+
+    fun toCordappVersionInfo() : CordappVersionInfo {
+        return CordappVersionInfo(cordappGroup!!, cordappName!!, cordappVersion!!, lastUpdated!!)
     }
+}
 
+/**
+ * Allows BNO to store and to retrieve reported CorDapp versions from the database
+ */
+@CordaService
+class DatabaseService(private val serviceHub : AppServiceHub) : SingletonSerializeAsToken() {
     /**
      * Updates or inserts information about a cordapp version
      */
     fun updateCordappVersionInfo(party : Party, cordappVersionInfo : CordappVersionInfo) {
-        val session : Connection = serviceHub.jdbcSession()
-
         // trying update first
-        val updated : Int = tryUpdateCordappVersionInfo(session, party, cordappVersionInfo)
+        val updated : Int = tryUpdateCordappVersionInfo(party, cordappVersionInfo)
 
         // if nothing has been updated then insert
-        if (updated == 0) insertCordappVersionInfo(session, party, cordappVersionInfo)
+        if (updated == 0) insertCordappVersionInfo(party, cordappVersionInfo)
     }
 
-    private fun tryUpdateCordappVersionInfo(session : Connection, party : Party, cordappVersionInfo : CordappVersionInfo) : Int {
-        val updateQuery = """
-            update $CORDAPP_VERSION_INFO_TABLE
-            set $CORDAPP_VERSION = ?, $LAST_UPDATED = ?
-            where $PARTY = ? AND $CORDAPP_GROUP = ? AND $CORDAPP_NAME = ?
+    private fun tryUpdateCordappVersionInfo(party : Party, cordappVersionInfo : CordappVersionInfo) : Int {
+        return serviceHub.withEntityManager {
+            val updateQuery = """
```

Comment:
Should this use `CriteriaBuilder` (i.e. `criteriaBuilder.createCriteriaUpdate(PersistentCordappVersionInfo::class.java)`...) instead of a native query?

Reply:
This is actually HQL :) It gets converted to a DB-specific syntax by hibernate.